src/coreclr/src/gc/gc.cpp

```diff
@@ -4016,7 +4022,7 @@ BOOL gc_heap::reserve_initial_memory (size_t normal_size, size_t large_size, siz
 
     size_t requestedMemory = memory_details.block_count * (normal_size + large_size + pinned_size);
 
-    uint8_t* allatonce_block = (uint8_t*)virtual_alloc (requestedMemory, use_large_pages_p);
+    uint8_t* allatonce_block = pohExplicitlyZeroed ? nullptr : (uint8_t*)virtual_alloc (requestedMemory, use_large_pages_p);
```

Comment:
I would still allocate SOH/LOH in one block if POH is 0 - the whole reason is we want them to be allocated together as much as possible. if we can we should still make them in one block. 

also we should only do a separate allocation for POH if it's 0 and if we are using large pages. 

something like this (I just wrote this quickly; you should definitely verify):

```c++
    bool seperated_poh_p = (pohExplicitlyZeroed && use_large_pages_p);
    size_t temp_pinned_size = (seperated_poh_p ? 0 : pinned_size);
    size_t requestedMemory = memory_details.block_count * (normal_size + large_size + temp_pinned_size);

    uint8_t* allatonce_block = (uint8_t*)virtual_alloc (requestedMemory, use_large_pages_p);
    if (allatonce_block)
    {
        g_gc_lowest_address = allatonce_block;
        g_gc_highest_address = allatonce_block + requestedMemory;
        memory_details.allocation_pattern = initial_memory_details::ALLATONCE;

        for (int i = 0; i < memory_details.block_count; i++)
        {
            memory_details.initial_normal_heap[i].memory_base = allatonce_block + 
                                                             (i * normal_size);
            memory_details.initial_large_heap[i].memory_base = allatonce_block +
                (memory_details.block_count * normal_size) + (i * large_size);
            memory_details.initial_pinned_heap[i].memory_base = allatonce_block +
                (memory_details.block_count * (normal_size + large_size)) + (i * temp_pinned_size);
        }

        if (seperated_poh_p)
        {
            uint8_t* poh_block = (uint8_t*)virtual_alloc ((memory_details.block_count * pinned_size), false);
            for (int i = 0; i < memory_details.block_count; i++)
            {
                memory_details.initial_pinned_heap[i].memory_base = allatonce_block +
                    (memory_details.block_count * (normal_size + large_size)) + (i * pinned_size);
            }

            return TRUE;
        }
    }
```

Reply:
this is not correct 
` g_gc_highest_address = allatonce_block + requestedMemory;`

should include the pinned part as well if it's allocated separately.